src/goto-programs/string_instrumentation.cpp

```diff
@@ -610,6 +649,199 @@ void string_instrumentationt::do_strncmp(
+  program.add(goto_programt::make_decl(content_symbol_expr));
+  program.add(goto_programt::make_assignment(content_symbol_expr, char_ptr));
+
+  // char (*nondet_infinite_array_ponter)[\infty];
+  const symbolt nondet_infinite_array_pointer = new_aux_symbol(
+    "cprover_string_index_of_func::nondet_infinite_array_pointer",
+    pointer_type(array_typet{char_type(), infinity_exprt(index_type())}),
+    symbol_table);
+  const symbol_exprt nondet_infinite_array_pointer_expr =
+    nondet_infinite_array_pointer.symbol_expr();
+  program.add(goto_programt::make_decl(nondet_infinite_array_pointer_expr));
+  program.add(goto_programt::make_assignment(
+    nondet_infinite_array_pointer_expr,
+    typecast_exprt{address_of_exprt{char_ptr},
+                   nondet_infinite_array_pointer_expr.type()}));
+
+  // cprover_associate_length_to_array_func(nondet_infinite_array_pointer,
+  //                                        string_length);
+  const auto refined_string_expr = refined_string_exprt{
+    length_symbol_expr,
+    content_symbol_expr,
+    refined_string_typet{index_type(), to_pointer_type(char_ptr.type())}};
+  const symbolt return_array_length = new_aux_symbol(
+    "cprover_string_index_of_func::return_array_length",
+    index_type(),
+    symbol_table);
+  dereference_exprt nondet_array_expr{nondet_infinite_array_pointer_expr};
+  const auto cprover_associate_length_to_array_func_symbol = symbol_exprt{
+    ID_cprover_associate_length_to_array_func,
+    mathematical_function_typet{{nondet_array_expr.type(), length_symbol.type},
+                                return_array_length.type}};
+  const auto apply_associate_length_to_array = function_application_exprt{
+    cprover_associate_length_to_array_func_symbol,
+    {nondet_array_expr, refined_string_expr.length()},
+    index_type()};
+  const auto return_length_expr = return_array_length.symbol_expr();
+  program.add(goto_programt::make_decl(return_length_expr));
+  program.add(goto_programt::make_assignment(
+    return_length_expr, apply_associate_length_to_array));
+
+  // cprover_associate_array_to_pointer_func(nondet_infinite_array_pointer,
+  //                                         src);
+  const address_of_exprt array_pointer(
+    index_exprt(nondet_array_expr, from_integer(0, index_type())));
+  program.add(goto_programt::make_assignment(array_pointer, char_ptr));
+  const symbolt return_sym_pointer =
+    new_aux_symbol("return_array_pointer", index_type(), symbol_table);
+  const auto cprover_associate_array_to_pointer_func_symbol =
+    symbol_exprt{ID_cprover_associate_array_to_pointer_func,
+                 mathematical_function_typet{
+                   {nondet_array_expr.type(), array_pointer.type()},
+                   return_sym_pointer.type}};
+  const auto apply_associate_pointer_to_array =
+    function_application_exprt{cprover_associate_array_to_pointer_func_symbol,
+                               {nondet_array_expr, char_ptr}};
```

Comment:
:bulb: These concepts I assume are shared with other string stuff, perhaps we could extract a utility for producing these exprs?